src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -626,6 +626,11 @@ inferDataConstructor
   -> DataConstructorDeclaration
   -> m (DataConstructorDeclaration, SourceType)
 inferDataConstructor tyCtor DataConstructorDeclaration{..} = do
+  -- Check that type synonyms can expand--we want partially applied type
+  -- synonyms to raise that error instead of KindsDoNotUnify--but keep the
+  -- unexpanded types in the declaration so that things like the transitive
+  -- exports check still work.
```

Comment:
I'm not saying you should change this now, but by this reasoning, should transitive export checks be happening before type-checking?

Reply:
Ideally, I think that would work? I don't think the environment prior to type checking has all of the entries that the transitive export checks use, but maybe things could be moved around to fix that.